pyqtgraph/functions.py

```diff
@@ -1752,51 +1750,120 @@ def arrayToQPath(x, y, connect='all'):
             if first < len(x):
                 # Replace all non-finite entries from beginning of arr with the first finite one
                 idx[:first] = first
-                arr[1:-1] = arr[1:-1][idx]
+                arr[:] = arr[:][idx]
+
 
     # decide which points are connected by lines
     if eq(connect, 'all'):
-        arr[1:-1]['c'] = 1
+        polygon = arrayToQPolygonF(
+            x,
+            y
+        )
+        path.addPolygon(polygon)
+        return path
     elif eq(connect, 'pairs'):
-        arr[1:-1]['c'][::2] = 0
-        arr[1:-1]['c'][1::2] = 1  # connect every 2nd point to every 1st one
+        xs = x.tolist()
+        ys = y.tolist()
+ 
+        lineTo = path.lineTo
+        moveTo = path.moveTo
+        for elementType, x, y in zip(
+            itertools.cycle((moveTo, lineTo)),
+            xs,
+            ys
+        ):
+            elementType(x, y)
+        return path
     elif eq(connect, 'finite'):
         # Let's call a point with either x or y being nan is an invalid point.
         # A point will anyway not connect to an invalid point regardless of the
         # 'c' value of the invalid point. Therefore, we should set 'c' to 0 for
         # the next point of an invalid point.
         if isfinite is None:
             isfinite = np.isfinite(x) & np.isfinite(y)
-        arr[2:]['c'] = isfinite
+        arr[1:]['c'] = isfinite[:-1]
     elif isinstance(connect, np.ndarray):
-        arr[2:-1]['c'] = connect[:-1]
+        arr[1:]['c'] = connect[:-1]
     else:
         raise Exception('connect argument must be "all", "pairs", "finite", or array')
 
-    arr[1]['c'] = 0  # the first vertex has no previous vertex to connect
+    arr[0]['c'] = 0  # the first vertex has no previous vertex to connect
 
-    byteview.data[-20:-16] = struct.pack('>i', 0)  # cStart
-    byteview.data[-16:-12] = struct.pack('>i', 0)  # fillRule (Qt.OddEvenFill)
+    # cStart, fillRule (Qt.OddEvenFill)
+    struct.pack_into('>ii', backstore, 4+n*20, 0, 0)
 
     # create datastream object and stream into path
 
     ## Avoiding this method because QByteArray(str) leaks memory in PySide
     #buf = QtCore.QByteArray(arr.data[12:lastInd+4])  # I think one unnecessary copy happens here
 
-    path.strn = byteview.data[16:-12]  # make sure data doesn't run away
+    path.strn = backstore
     try:
         buf = QtCore.QByteArray.fromRawData(path.strn)
```

Comment:
The ```QtCore.QByteArray.fromRawData()``` code path is for PyQt5 and according to the C++ docs, it's supposed to be zero-copy, hence the ```path.strn = backstore``` line to hold a reference. However, I have tested that the PyQt5 implementation is _not_ zero-copy.

Therefore, we can simplify this whole chunk to:
```python
if QT_LIB == "PyQt6":
    buf = QtCore.QByteArray(path.strn, len(path.strn))
else:
    buf = QtCore.QByteArray(path.strn)
```
The special case for PyQt6 is because PyQt6 < 6.1.1 has a bug where the single argument constructor treats a NUL byte as a terminator.
https://www.riverbankcomputing.com/pipermail/pyqt/2021-May/043942.html

Reply:
Should we make the check be `if QT_LIB == "PyQt6" and Qt_Version < 6.1.1` ?

also not that I doubt you, but how are you checking if this is a zero copy operation or not?

Reply:
```python
In [1]: from PyQt5 import QtCore

In [2]: ba = bytearray(b'abcdef\0ghijkl')

In [3]: qba = QtCore.QByteArray.fromRawData(ba)

In [4]: qba
Out[4]: PyQt5.QtCore.QByteArray(b'abcdef\x00ghijkl')

In [5]: ba[-1] = ord(b'Z')

In [6]: ba
Out[6]: bytearray(b'abcdef\x00ghijkZ')

In [7]: qba
Out[7]: PyQt5.QtCore.QByteArray(b'abcdef\x00ghijkl')
```

Reply:
>Should we make the check be if QT_LIB == "PyQt6" and Qt_Version < 6.1.1 ?

That would be good for documenting the bug. But for now, I am just guessing that the fix will land in PyQt6 6.1.1.
The two argument version always works. So maybe add a comment instead?

Reply:
Looks like I was wrong. In PyQt6 6.1.1, the two argument version no longer works, so will need version checking.